### ISSUE TITLE
WebHostService does not log service state changes

### ISSUE THREAD

**Issue by Erikvl87** [aspnet/AspNetCore#12352]
### Describe the bug
The [WebHostService](https://github.com/aspnet/AspNetCore/blob/v2.2.5/src/Hosting/WindowsServices/src/WebHostService.cs) implements ServiceBase which has a public property called  `AutoLog` which is documented as follows: [Indicates whether to report Start, Stop, Pause, and Continue commands in the event](https://github.com/dotnet/corefx/blob/v2.2.5/src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L83). This doesn't seem to work for the application (`netcoreapp2.2`) I am developing. I am following the guide [Host ASP.NET Core in a Windows Service](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-2.2&tabs=visual-studio) to accomplish this. One of my requirements is that I log state changes to the event viewer.

The documentation shows how to run as a custom service so I can [handle starting and stopping events](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-2.2&tabs=visual-studio#handle-starting-and-stopping-events) myself, but my understanding is that these event logs should already been covered by the [ServiceBase](https://github.com/dotnet/corefx/blob/v2.2.5/src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs) using the `AutoLog` property. I cannot find any logs of the service being started or stopped in the event viewer (running Windows 10).

I have managed to accomplish this by creating a `CustomWebHostService`, but should I? Is the documentation / implementation of AutoLog on the ServiceBase (or any other relevant code) bugged?

### To Reproduce
Follow the guide [Host ASP.NET Core in a Windows Service](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-2.2&tabs=visual-studio). Install and start the service and observe that the event viewer does not contain any start/stop/pause events.

**Comment by anurse** [comment]
[`AutoLog` uses an EventSource named after `ServiceName`](https://docs.microsoft.com/en-us/dotnet/api/system.serviceprocess.servicebase.autolog?view=netframework-4.8) value. Are you running as an admin or have you previously registered that source? If you're using a custom source name (as `AutoLog` does), the name has to first be registered by a user with Admin permissions. If your service is running as an Admin, it will be auto-registered, but if not you need to manually register it by running as Admin and calling an API like [`EventLog.CreateEventSource`](https://docs.microsoft.com/en-us/dotnet/api/system.diagnostics.eventlog.createeventsource?view=netframework-4.8)

**Comment by anurse** [comment]
> **Edit:** I did find that Windows Server 2019 does log start/stopped events under id 7036. I guess this somehow does respect the `AutoLog` property.

Do you mean that it only works on Windows Server 2019 but still isn't working on other OS versions?

**Comment by Erikvl87** [comment]
> Do you mean that it only works on Windows Server 2019 but still isn't working on other OS versions?

Yes, that is what I ment. In Windows Server 2019 the `Service Control Manager`  is logging eventId `7036` which describes that `The <service name> service entered the <running/stopped> state.` 
```
<Event xmlns="http://schemas.microsoft.com/win/2004/08/events/event">
	<System>
		<Provider Name="Service Control Manager" Guid="{555908d1-a6d7-4695-8e1e-26931d2012f4}" EventSourceName="Service Control Manager" />
		<EventID Qualifiers="16384">7036</EventID>
		<Version>0</Version>
		<Level>4</Level>
		<Task>0</Task>
		<Opcode>0</Opcode>
		<Keywords>0x8080000000000000</Keywords>
		<TimeCreated SystemTime="2019-07-19T13:34:31.812574400Z" />
		<EventRecordID>605</EventRecordID>
		<Correlation />
		<Execution ProcessID="696" ThreadID="4928" />
		<Channel>System</Channel>
		<Computer>MY_SERVER</Computer>
		<Security />
	</System>
	<EventData>
		<Data Name="param1">MY_APP</Data>
		<Data Name="param2">running</Data>
		<Binary>46006F006E0074006F005200650076006900650077002F0034000000</Binary>
	</EventData>
</Event>
```

I did not get this on Windows 10. I was assuming that these events are created somewhere internally based on the [AutoLog](https://github.com/dotnet/corefx/blob/v2.2.5/src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L83) property since its documentation is literally `Indicates whether to report Start, Stop, Pause, and Continue commands in the event` (log). But I think my assumption is incorrect, since the documentation of AutoLog that you mentioned states that it uses the **Application** event log while I am looking at logs from the **Service Control Manager** .

>  If you're using a custom source name (as AutoLog does), the name has to first be registered by a user with Admin permissions. If your service is running as an Admin, it will be auto-registered, but if not you need to manually register it by running as Admin and calling an API like EventLog.CreateEventSource

I followed the guide [Service user account](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-2.2&tabs=visual-studio#service-user-account) to create the `Service user account`, `Log on as a service rights` and `Create and manage the Windows Service`.

I just changed the permissions of the concerning account to be part of the Administrators group, but still no service start/stopped events from the `Service Control Manager` on my Windows 10 machine, nor from the `Application` event source.

> [`AutoLog` uses an EventSource named after `ServiceName`](https://docs.microsoft.com/en-us/dotnet/api/system.serviceprocess.servicebase.autolog?view=netframework-4.8) value. Are you running as an admin or have you previously registered that source?

Just to make sure we are on the same page; I was expecting event logs (like the one above, 7036) for service started/stopped events without the need of [AddEventLog](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.eventloggerfactoryextensions.addeventlog?view=aspnetcore-2.2), nor following the guide on how to [Handle starting and stopping events](https://docs.microsoft.com/en-us/aspnet/core/host-and-deploy/windows-service?view=aspnetcore-2.2&tabs=visual-studio#handle-starting-and-stopping-events) nor applying any custom configuration for the event viewer whatsoever. And that expectation was indeed met on Windows Server 2019 (by the logs from the `Service Control Manager`) but not on our Windows 10 development machines.

I was expecting the  `WebHostService : ServiceBase` to [log](https://github.com/dotnet/corefx/blob/v2.2.5/src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L840) that the [Service started successfully](https://github.com/dotnet/corefx/blob/c390ce7df50252e11f5d322276e9d19e046d1332/src/System.ServiceProcess.ServiceController/src/Resources/Strings.resx#L121).

So basically I have an app that uses the default [RunAsService](https://github.com/aspnet/AspNetCore/blob/v2.2.5/src/Hosting/WindowsServices/src/WebHostWindowsServiceExtensions.cs#L36) method, which creates a WebHostService and [run](https://github.com/dotnet/corefx/blob/v2.2.5/src/System.ServiceProcess.ServiceController/src/System/ServiceProcess/ServiceBase.cs#L598)s that service. Here is a simplified version of my code:

```
public static void Main(string[] args)
{
	var pathToExe = Process.GetCurrentProcess().MainModule?.FileName;
	var pathToContentRoot = Path.GetDirectoryName(pathToExe);
	Directory.SetCurrentDirectory(pathToContentRoot);
	var host = CreateWebHostBuilder().Build();
	host.RunAsService();
}

private static IWebHostBuilder CreateWebHostBuilder()
{
	var builder = new WebHostBuilder()
		.UseContentRoot(Directory.GetCurrentDirectory())
		.ConfigureAppConfiguration((hostingContext, config) =>
			config.AddEnvironmentVariables())
		.UseStartup<Startup>()
		.UseKestrel();
	return builder;
}
```

**Comment by anurse** [comment]
> Just to make sure we are on the same page; I was expecting event logs (like the one above, 7036) for service started/stopped events without the need of [AddEventLog](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.eventloggerfactoryextensions.addeventlog?view=aspnetcore-2.2)

Yep, we're on the same page. That's what I'm talking about, the built-in logging to EventLog provided by `ServiceBase`.

Honestly, we haven't really validated what `AutoLog` does in the `WebHostService` scenario because our general recommendation has been to just use the built-in logging functionality in ASP.NET Core (since the host should be logging lifetime events like start and stop to `ILogger` automatically). For example, `WebHostService` never sets `ServiceName` which may affect how `AutoLog` works.

**Comment by anurse** [comment]
I think the best thing here to help move forward is if you can provide a runnable sample that reproduces the problem. That way we can see the whole project and get the full context.

**Comment by Erikvl87** [comment]
I've created a runnable sample at [https://github.com/Erikvl87/AspNetCore-issue-12352](https://github.com/Erikvl87/AspNetCore-issue-12352).

And while doing so, I've found that you are right: The `ServiceName` is not being set and that is the cause of the logs not being written. After hard-coding it to `Application` everything works. I've added the workaround to the sample as well.

**Comment by anurse** [comment]
Setting the `ServiceName` to `Application` isn't really appropriate. Maybe we could set it to the assembly name.

**Comment by Erikvl87** [comment]
As you've told me the AutoLog uses an EventSource named after ServiceName. That would mean that the assembly name first has to be registered by a user with Admin permissions. Would that be sensible default behavior?

I reproduced this in my sample repository (just changing the ServiceName to something else than `application` and verified that I do not see the service start/stopped logs anymore.

**Comment by anurse** [comment]
> So that would mean that the assembly name first has to be registered by a user with Admin permissions.

Yep, either you need to run the app itself as an Admin (not generally desirable) or run some other application/tool as an Admin to register the source (I don't have a specific suggestion for one, if one exists).

> Maybe add configuration options to be able to explicitly set the name when using the [RunAsService()](https://github.com/aspnet/AspNetCore/blob/v2.2.5/src/Hosting/WindowsServices/src/WebHostWindowsServiceExtensions.cs#L36) [](/aspnet/AspNetCore/blob/HEAD@{2019-07-30T07:16:27Z}/src/Hosting/WindowsServices/src/WebHostWindowsServiceExtensions.cs#L36) extension method instead.

That's a pretty reasonable suggestion. Honestly, an API like this would seem OK to me:

```csharp
public static class WebHostWindowsServiceExtensions
{
    // Exiting method
    public static void RunAsService(this IWebHost host);

    // New overload
    public static void RunAsService(this IWebHost host, string serviceName);
}
```

Defaulting the service name to the assembly name makes sense, but so too does allowing the app to override that. I suspect we'd take a PR with that change (or we can consider it while planning for future releases), though I'd need some other folks on the team to take a look and see what they thing (cc @Tratcher @davidfowl @rynowak).

**Comment by Tratcher** [comment]
Note this functionality has moved to the generic IHost and any improvement should be made there instead. (And this issue should be moved to Extensions).

There were some design changes in the move and RunAsService no longer exists there. We would use the normal options pattern to provide a name.
https://github.com/aspnet/Extensions/blob/master/src/Hosting/WindowsServices/src/WindowsServiceLifetimeHostBuilderExtensions.cs
